src/Sarif.Multitool/Rules/SARIF2010.ProvideCodeSnippets.cs

```diff
@@ -65,5 +78,46 @@ private void AnalyzeRegion(Region region, string regionPointer)
                     nameof(RuleResources.SARIF2010_ProvideCodeSnippets_Note_Default_Text));
             }
         }
+
+        private bool AnalyzeArtifactLocation(ArtifactLocation artifactLocation)
+        {
+            // no artifactLocation / no artifacts, so we should look for the snippet
+            if (artifactLocation == null || this.artifacts == null)
+            {
+                return true;
+            }
+
+            artifactLocation.TryReconstructAbsoluteUri(this.originalUriBaseIds, out Uri resolvedUri);
```

Comment:
>TryReconstructAbsoluteUri [](start = 29, length = 25)

`TryReconstructAbsoluteUri` can return `false`, in which case `resolvedUri` will be null, and you'll need to use `artifactLocation.Uri` instead. So, something like this:
```
Uri resolvedUri;
if (!artifactLocation.TryReconstructAbsoluteUri(this.originalUriBaseIds, out resolvedUri))
{
    resolvedUri = artifactLocation.Uri;
}
```

Please look at `TryReconstructAbsoluteUri` to see how it can fail; then provide a test case that exercises this logic.

Reply:
Come to think of it:

```
if (!artifactLocation.TryReconstructAbsoluteUri(this.originalUriBaseIds, out Uri resolvedUri))
{
    return true;
}
```

If a SARIF viewer can't resolve the URI, it won't be able to find the artifact, so it should insist on the snippet being present.

We still need a test case for this, thought.

---
In reply to: [462536376](https://github.com/microsoft/sarif-sdk/pull/2019#discussion_r462536376) [](ancestors = 462536376)